stratum/hal/lib/barefoot/bf_sde_wrapper.cc

```diff
@@ -309,6 +319,785 @@ ::util::Status BfSdeWrapper::SetTmCpuPort(int device, int port) {
+  if (VLOG_IS_ON(2)) {
+    auto real_session = std::dynamic_pointer_cast<Session>(session);
+    CHECK_RETURN_IF_FALSE(real_session);
+
+    auto bf_dev_tgt = GetDeviceTarget(device);
+    const bfrt::BfRtTable* table;
+
+    // Dump group table
+    LOG(INFO) << "#### $pre.mgid ####";
+    RETURN_IF_BFRT_ERROR(
+        bfrt_info_->bfrtTableFromNameGet(kPreMgidTable, &table));
+    std::vector<std::unique_ptr<bfrt::BfRtTableKey>> keys;
+    std::vector<std::unique_ptr<bfrt::BfRtTableData>> datums;
+    RETURN_IF_BFRT_ERROR(
+        bfrt_info_->bfrtTableFromNameGet(kPreMgidTable, &table));
+    RETURN_IF_ERROR(GetAllEntries(real_session->bfrt_session_, bf_dev_tgt,
+                                  table, &keys, &datums));
+    for (size_t i = 0; i < keys.size(); ++i) {
+      const std::unique_ptr<bfrt::BfRtTableData>& table_data = datums[i];
+      const std::unique_ptr<bfrt::BfRtTableKey>& table_key = keys[i];
+      PrintMcGroupEntry(table, table_key.get(), table_data.get());
+    }
+    LOG(INFO) << "###################";
+
+    // Dump node table
+    LOG(INFO) << "#### $pre.node ####";
+    RETURN_IF_BFRT_ERROR(
+        bfrt_info_->bfrtTableFromNameGet(kPreNodeTable, &table));
+    RETURN_IF_ERROR(GetAllEntries(real_session->bfrt_session_, bf_dev_tgt,
+                                  table, &keys, &datums));
+    for (size_t i = 0; i < keys.size(); ++i) {
+      const std::unique_ptr<bfrt::BfRtTableData>& table_data = datums[i];
+      const std::unique_ptr<bfrt::BfRtTableKey>& table_key = keys[i];
+      PrintMcNodeEntry(table, table_key.get(), table_data.get());
+    }
+    LOG(INFO) << "###################";
+  }
+  return ::util::OkStatus();
+}
+
+::util::StatusOr<uint32> BfSdeWrapper::GetFreeMulticastNodeId(
+    int device, std::shared_ptr<BfSdeInterface::SessionInterface> session) {
+  auto real_session = std::dynamic_pointer_cast<Session>(session);
+  CHECK_RETURN_IF_FALSE(real_session);
+
+  auto bf_dev_tgt = GetDeviceTarget(device);
+  const bfrt::BfRtTable* table;
+  RETURN_IF_BFRT_ERROR(bfrt_info_->bfrtTableFromNameGet(kPreNodeTable, &table));
+  size_t table_size;
+  RETURN_IF_BFRT_ERROR(table->tableSizeGet(&table_size));
+  uint32 usage;
+  RETURN_IF_BFRT_ERROR(table->tableUsageGet(
+      *real_session->bfrt_session_, bf_dev_tgt,
+      bfrt::BfRtTable::BfRtTableGetFlag::GET_FROM_SW, &usage));
+  std::unique_ptr<bfrt::BfRtTableKey> table_key;
+  std::unique_ptr<bfrt::BfRtTableData> table_data;
+  RETURN_IF_BFRT_ERROR(table->keyAllocate(&table_key));
+  RETURN_IF_BFRT_ERROR(table->dataAllocate(&table_data));
+  uint32 id = usage;
+  for (size_t _ = 0; _ < table_size; ++_) {
+    // Key: $MULTICAST_NODE_ID
+    RETURN_IF_ERROR(SetField(table_key.get(), kMcNodeId, id));
+    bf_status_t status = table->tableEntryGet(
+        *real_session->bfrt_session_, bf_dev_tgt, *table_key,
+        bfrt::BfRtTable::BfRtTableGetFlag::GET_FROM_SW, table_data.get());
+    if (status == BF_OBJECT_NOT_FOUND) {
+      return id;
```

Comment:
Is it possible that multiple threads calling this method and get the same ID?
Since we are using the reader lock

Reply:
yes.

I initially though about changing this function to `AllocateFree...` and reserve the ID right here. Later users would have to call modify, though. But even that would not guarantee uniqueness, just make it less likely.

Alternatives I see:

- BfRt Transactions, would require changing the batching logic. Can't have a transaction inside a batch.
- Exclusive locking, like you noted
- Busy trying. Basically retry the insert with a new ID when a EXISTS error is returned. At some point there should be no other threads.

Happy to hear other ideas.